### PR TITLE
Restore no-crash behaviour when read-only config file(s)

### LINKED ISSUE #8301: Crash when paper-global.yml and settings directory are read-only
### Stack trace

```
[18:01:07 ERROR]: Encountered an unexpected exception
org.spongepowered.configurate.ConfigurateException: []: java.nio.file.AccessDeniedException: /mcconfig/.22351533673970431302813050paper-global.yml.tmp
        at org.spongepowered.configurate.loader.AbstractConfigurationLoader.save(AbstractConfigurationLoader.java:205) ~[configurate-core-4.1.2.jar:?]
        at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:114) ~[paper-1.19.2.jar:git-Paper-130]
        at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:97) ~[paper-1.19.2.jar:git-Paper-130]
        at io.papermc.paper.configuration.PaperConfigurations.initializeGlobalConfiguration(PaperConfigurations.java:180) ~[paper-1.19.2.jar:git-Paper-130]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:214) ~[paper-1.19.2.jar:git-Paper-130]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1125) ~[paper-1.19.2.jar:git-Paper-130]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305) ~[paper-1.19.2.jar:git-Paper-130]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.nio.file.AccessDeniedException: /mcconfig/.22351533673970431302813050paper-global.yml.tmp
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:90) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
        at sun.nio.fs.UnixCopyFile.copyFile(UnixCopyFile.java:246) ~[?:?]
        at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:603) ~[?:?]
        at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258) ~[?:?]
        at java.nio.file.Files.copy(Files.java:1305) ~[?:?]
        at org.spongepowered.configurate.loader.AtomicFiles.atomicBufferedWriter(AtomicFiles.java:85) ~[configurate-core-4.1.2.jar:?]
        at org.spongepowered.configurate.loader.AtomicFiles.lambda$atomicWriterFactory$0(AtomicFiles.java:58) ~[configurate-core-4.1.2.jar:?]
        at org.spongepowered.configurate.loader.AbstractConfigurationLoader.save(AbstractConfigurationLoader.java:188) ~[configurate-core-4.1.2.jar:?]
        ... 7 more
[18:01:07 ERROR]:       Cause of unexpected exception was
java.nio.file.AccessDeniedException: /mcconfig/.22351533673970431302813050paper-global.yml.tmp
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:90) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
        at sun.nio.fs.UnixCopyFile.copyFile(UnixCopyFile.java:246) ~[?:?]
        at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:603) ~[?:?]
        at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258) ~[?:?]
        at java.nio.file.Files.copy(Files.java:1305) ~[?:?]
        at org.spongepowered.configurate.loader.AtomicFiles.atomicBufferedWriter(AtomicFiles.java:85) ~[configurate-core-4.1.2.jar:?]
        at org.spongepowered.configurate.loader.AtomicFiles.lambda$atomicWriterFactory$0(AtomicFiles.java:58) ~[configurate-core-4.1.2.jar:?]
        at org.spongepowered.configurate.loader.AbstractConfigurationLoader.save(AbstractConfigurationLoader.java:188) ~[configurate-core-4.1.2.jar:?]
        at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:114) ~[paper-1.19.2.jar:git-Paper-130]
        at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:97) ~[paper-1.19.2.jar:git-Paper-130]
        at io.papermc.paper.configuration.PaperConfigurations.initializeGlobalConfiguration(PaperConfigurations.java:180) ~[paper-1.19.2.jar:git-Paper-130]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:214) ~[paper-1.19.2.jar:git-Paper-130]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1125) ~[paper-1.19.2.jar:git-Paper-130]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305) ~[paper-1.19.2.jar:git-Paper-130]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

### Plugin and Datapack List

N/A

### Actions to reproduce (if known)

1. Populate a directory (eg `my_configs/`) with a preset `paper-global.yml` and `paper-world-defaults.yml` files
2. Set directory and files as read-only (or change ownership to some other user and remove the writable flag)
3. Start paper with argument `--paper-settings-directory my_configs`

### Paper version

[18:01:07 INFO]: This server is running Paper version git-Paper-130 (MC: 1.19.2)

### Other

The use case is to have versioned configuration files that are immutable once applied to the server. Paper should not crash if the configuration files are present and correct, but simply read-only, unless there is a misconfiguration inside or a necessary migration.

IMO this is a regression as Paper did not crash in this situation (but definitely complained) prior to migrating the configuration handling to Configurate.

### PR DESCRIPTION
**Pull request by karthanistyr**
Prior to migrating handling configuration with Configurate,
paper did not crash when it found it couldn't overwrite the
existing paper.yml configuration file (but left a stack trace).

After the migration, paper crashes when it is unable to overwrite
any of paper-global.yml, paper-world-defaults.yml or a
world-specific config file.

This aims at returning to the previous behaviour of not
crashing when paper is unable to overwrite a configuration
file.

Signed-off-by: Antoine Mazeas <antoine@karthanis.net>

Fix #8301

### CODE REVIEW

**Comment by karthanistyr** [comment]
Trace in console:
```
servercliffs_1  | [10:23:10 WARN]: Can not write to file server.properties, skipping.
[...]
servercliffs_1  | [10:23:14 WARN]: [io.papermc.paper.configuration.Configurations] Could not save /mcconfig/paper-global.yml: Paper could not persist the full set of configuration settings in the configuration file. Any setting missing from the configuration file will be set with its default value in memory. Admins should make sure to review the configuration documentation at https://docs.papermc.io/paper/configuration for more details.
servercliffs_1  | [10:23:14 WARN]: [io.papermc.paper.configuration.Configurations] Exception: {}
servercliffs_1  | org.spongepowered.configurate.ConfigurateException: []: java.nio.file.AccessDeniedException: /mcconfig/.369675404516007-462797823paper-global.yml.tmp
servercliffs_1  | 	at org.spongepowered.configurate.loader.AbstractConfigurationLoader.save(AbstractConfigurationLoader.java:205) ~[configurate-core-4.1.2.jar:?]
servercliffs_1  | 	at io.papermc.paper.configuration.Configurations.trySaveFileNode(Configurations.java:103) ~[paper-1.19.2.jar:git-Paper-"168db18"]
servercliffs_1  | 	at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:126) ~[paper-1.19.2.jar:git-Paper-"168db18"]
servercliffs_1  | 	at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:98) ~[paper-1.19.2.jar:git-Paper-"168db18"]
servercliffs_1  | 	at io.papermc.paper.configuration.PaperConfigurations.initializeGlobalConfiguration(PaperConfigurations.java:180) ~[paper-1.19.2.jar:git-Paper-"168db18"]
servercliffs_1  | 	at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:214) ~[paper-1.19.2.jar:git-Paper-"168db18"]
servercliffs_1  | 	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1126) ~[paper-1.19.2.jar:git-Paper-"168db18"]
servercliffs_1  | 	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305) ~[paper-1.19.2.jar:git-Paper-"168db18"]
servercliffs_1  | 	at java.lang.Thread.run(Thread.java:833) ~[?:?]
servercliffs_1  | Caused by: java.nio.file.AccessDeniedException: /mcconfig/.369675404516007-462797823paper-global.yml.tmp
servercliffs_1  | 	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:90) ~[?:?]
servercliffs_1  | 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106) ~[?:?]
servercliffs_1  | 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
servercliffs_1  | 	at sun.nio.fs.UnixCopyFile.copyFile(UnixCopyFile.java:246) ~[?:?]
servercliffs_1  | 	at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:603) ~[?:?]
servercliffs_1  | 	at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258) ~[?:?]
servercliffs_1  | 	at java.nio.file.Files.copy(Files.java:1305) ~[?:?]
servercliffs_1  | 	at org.spongepowered.configurate.loader.AtomicFiles.atomicBufferedWriter(AtomicFiles.java:85) ~[configurate-core-4.1.2.jar:?]
servercliffs_1  | 	at org.spongepowered.configurate.loader.AtomicFiles.lambda$atomicWriterFactory$0(AtomicFiles.java:58) ~[configurate-core-4.1.2.jar:?]
servercliffs_1  | 	at org.spongepowered.configurate.loader.AbstractConfigurationLoader.save(AbstractConfigurationLoader.java:188) ~[configurate-core-4.1.2.jar:?]
servercliffs_1  | 	... 8 more
```

**Comment by karthanistyr** [comment]
This is like #8306 with addressed comments, for your consideration. This could be a satisfactory stop-gap while a broader design is decided by the core team.

File: patches/server/0004-Paper-config-files.patch
Comment (Machine-Maker) [inline patches/server/0004-Paper-config-files.patch]: Paper comment

File: patches/server/0004-Paper-config-files.patch
Comment (Machine-Maker) [inline patches/server/0004-Paper-config-files.patch]: You can use imports here, as this patch is the patch that creates the file. So go ahead and import LogUtils and Logger and use those directly.

We like to avoid adding imports to files that we are changing as that can create useless conflicts on rebases. But if the file is being first created, you can add as many imports as you want as its never going to conflict on a rebase.

File: patches/server/0004-Paper-config-files.patch
Comment (Machine-Maker) [inline patches/server/0004-Paper-config-files.patch]: add import

**Review by Machine-Maker — DISMISSED** [review]

File: patches/server/0004-Paper-config-files.patch
Comment (Lulu13022002) [inline patches/server/0004-Paper-config-files.patch]: The {} is not interpreted (see your log) because of the type. A lot of logger supports an optional parameter at the end of their function to print directly the stacktrace. In short you can just remove the " {}" at the end and it will still work
But i think you can remove this line totally because we already know that it's an exception no need to write Exception: ...

Comment (karthanistyr) [reply patches/server/0004-Paper-config-files.patch]: yes, this was to benefit from `Logger.warn(String, Throwable)` but apparently this can be compressed into a single call, ref: https://www.slf4j.org/faq.html#paramException

Wouldn't we want to keep logging the stack trace to be consistent with other skips like with spigot.yml ?

Comment (Lulu13022002) [reply patches/server/0004-Paper-config-files.patch]: Yes you can compress that into a single call, also i agree with the consistency between the other files. Basically when this is a real permission issue (and not a wanted behavior) the stacktrace help a lot

File: patches/server/0004-Paper-config-files.patch
Comment (karthanistyr) [reply patches/server/0004-Paper-config-files.patch]: Done

File: patches/server/0004-Paper-config-files.patch
Comment (karthanistyr) [reply patches/server/0004-Paper-config-files.patch]: Done

File: patches/server/0004-Paper-config-files.patch
Comment (karthanistyr) [reply patches/server/0004-Paper-config-files.patch]: Done as mentioned

File: patches/server/0004-Paper-config-files.patch
Comment (karthanistyr) [reply patches/server/0004-Paper-config-files.patch]: Done

**Comment by karthanistyr** [comment]
New trace format:
```
[19:12:38 WARN]: [io.papermc.paper.configuration.Configurations] Could not save /mcconfig/paper-global.yml: Paper could not persist the full set of configuration settings in the configuration file. Any setting missing from the configuration file will be set with its default value in memory. Admins should make sure to review the configuration documentation at https://docs.papermc.io/paper/configuration for more details.
org.spongepowered.configurate.ConfigurateException: []: java.nio.file.AccessDeniedException: /mcconfig/.574238817695928357832769paper-global.yml.tmp
	at org.spongepowered.configurate.loader.AbstractConfigurationLoader.save(AbstractConfigurationLoader.java:205) ~[configurate-core-4.1.2.jar:?]
	at io.papermc.paper.configuration.Configurations.trySaveFileNode(Configurations.java:106) ~[paper-1.19.2.jar:git-Paper-"5ab6209"]
	at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:128) ~[paper-1.19.2.jar:git-Paper-"5ab6209"]
	at io.papermc.paper.configuration.Configurations.initializeGlobalConfiguration(Configurations.java:101) ~[paper-1.19.2.jar:git-Paper-"5ab6209"]
	at io.papermc.paper.configuration.PaperConfigurations.initializeGlobalConfiguration(PaperConfigurations.java:180) ~[paper-1.19.2.jar:git-Paper-"5ab6209"]
	at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:214) ~[paper-1.19.2.jar:git-Paper-"5ab6209"]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1126) ~[paper-1.19.2.jar:git-Paper-"5ab6209"]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305) ~[paper-1.19.2.jar:git-Paper-"5ab6209"]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.nio.file.AccessDeniedException: /mcconfig/.574238817695928357832769paper-global.yml.tmp
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:90) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
	at sun.nio.fs.UnixCopyFile.copyFile(UnixCopyFile.java:246) ~[?:?]
	at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:603) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258) ~[?:?]
	at java.nio.file.Files.copy(Files.java:1305) ~[?:?]
	at org.spongepowered.configurate.loader.AtomicFiles.atomicBufferedWriter(AtomicFiles.java:85) ~[configurate-core-4.1.2.jar:?]
	at org.spongepowered.configurate.loader.AtomicFiles.lambda$atomicWriterFactory$0(AtomicFiles.java:58) ~[configurate-core-4.1.2.jar:?]
	at org.spongepowered.configurate.loader.AbstractConfigurationLoader.save(AbstractConfigurationLoader.java:188) ~[configurate-core-4.1.2.jar:?]
	... 8 more
```

File: patches/server/0004-Paper-config-files.patch
Comment (Lulu13022002) [reply patches/server/0004-Paper-config-files.patch]: The comment must be in the same line when it's only one line of diff

Comment (karthanistyr) [reply patches/server/0004-Paper-config-files.patch]: and done

**Comment by karthanistyr** [comment]
Is there anything else I must do before this gets re-reviewed and possibly merged ?

**Review by kennytv — APPROVED** [review]

**Comment by karthanistyr** [comment]
hopefully the conflicts are handled

**Comment by karthanistyr** [comment]
rebased

**Review by electronicboy — APPROVED** [review]

**Pull request merged**
